core/io.rs

```diff
@@ -0,0 +1,268 @@
+      }
+      BufView::Vec { vec, start, end } => buf_range(vec.deref(), *start, *end),
+    }
+  }
+}
+
+impl AsRef<[u8]> for BufView {
+  fn as_ref(&self) -> &[u8] {
+    self.deref()
+  }
+}
+
+impl From<ZeroCopyBuf> for BufView {
+  fn from(buf: ZeroCopyBuf) -> Self {
+    BufView::ZeroCopy {
+      buf,
+      start: 0,
+      end: None,
+    }
+  }
+}
+
+impl From<Vec<u8>> for BufView {
+  fn from(vec: Vec<u8>) -> Self {
+    BufView::Vec {
+      vec,
+      start: 0,
+      end: None,
+    }
+  }
+}
+
+impl From<bytes::Bytes> for BufView {
+  fn from(buf: bytes::Bytes) -> Self {
+    BufView::Bytes(buf)
+  }
+}
+
+impl From<BufView> for bytes::Bytes {
+  fn from(buf: BufView) -> Self {
+    match buf {
+      BufView::Empty => bytes::Bytes::new(),
+      BufView::Bytes(bytes) => bytes,
+      BufView::ZeroCopy { buf, start, end } => {
+        bytes::Bytes::from(buf_range(buf.deref(), start, end).to_vec())
+      }
+      BufView::Vec { vec, start, end } => {
+        let end = end.unwrap_or(vec.len());
+        bytes::Bytes::from(vec).slice(start..end)
+      }
+    }
+  }
+}
+
+/// BufMut is a wrapper around an underlying contiguous chunk of writable bytes.
+/// It can be created from a `ZeroCopyBuf` or a `Vec<u8>` and implements
+/// `DerefMut<[u8]>` and `AsMut<[u8]>`.
+///
+/// A `BufMut` can be turned into a `BufView` by calling `BufMut::into_view`.
+pub enum BufMut {
+  ZeroCopy(ZeroCopyBuf),
+  Vec(Vec<u8>),
+}
+
+impl BufMut {
+  pub fn new(len: usize) -> Self {
+    BufMut::Vec(vec![0; len])
+  }
+
+  /// Get the length of the underlying buffer.
+  pub fn len(&self) -> usize {
+    match self {
+      BufMut::ZeroCopy(buf) => buf.len(),
+      BufMut::Vec(vec) => vec.len(),
+    }
+  }
+
+  pub fn is_empty(&self) -> bool {
+    self.len() == 0
+  }
+
+  /// Turn this `BufMut` into a `BufView`.
+  pub fn into_view(self) -> BufView {
+    match self {
+      BufMut::ZeroCopy(buf) => BufView::ZeroCopy {
+        buf,
+        start: 0,
+        end: None,
+      },
+      BufMut::Vec(vec) => BufView::Vec {
+        vec,
+        start: 0,
+        end: None,
+      },
+    }
+  }
+}
+
+impl Deref for BufMut {
+  type Target = [u8];
+
+  fn deref(&self) -> &[u8] {
+    match self {
+      BufMut::ZeroCopy(buf) => buf.deref(),
+      BufMut::Vec(vec) => vec.deref(),
+    }
+  }
+}
+
+impl DerefMut for BufMut {
+  fn deref_mut(&mut self) -> &mut [u8] {
+    match self {
+      BufMut::ZeroCopy(buf) => buf.deref_mut(),
+      BufMut::Vec(vec) => vec.deref_mut(),
+    }
+  }
+}
+
+impl AsRef<[u8]> for BufMut {
+  fn as_ref(&self) -> &[u8] {
+    self.deref()
+  }
+}
+
+impl AsMut<[u8]> for BufMut {
+  fn as_mut(&mut self) -> &mut [u8] {
+    self.deref_mut()
+  }
+}
+
+impl From<ZeroCopyBuf> for BufMut {
+  fn from(buf: ZeroCopyBuf) -> Self {
+    BufMut::ZeroCopy(buf)
+  }
+}
+
+impl From<Vec<u8>> for BufMut {
+  fn from(buf: Vec<u8>) -> Self {
+    BufMut::Vec(buf)
+  }
+}
+
+pub enum WriteResponse {
```

Comment:
This name is confusing... at first glace I thought it's about writing HTTP response. How about `WriteEffect`?